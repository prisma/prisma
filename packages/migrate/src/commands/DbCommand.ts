import type { Command, Commands } from '@prisma/internals'
import { arg, format, HelpError, isError, unknownCommand } from '@prisma/internals'
import chalk from 'chalk'

export class DbCommand implements Command {
  public static new(cmds: Commands): DbCommand {
    return new DbCommand(cmds)
  }

  private static help = format(`
${process.platform === 'win32' ? '' : chalk.bold('üèãÔ∏è  ')}Manage your database schema and lifecycle during development.

${chalk.bold('Usage')}

  ${chalk.dim('$')} prisma db [command] [options]

${chalk.bold('Options')}

  -h, --help   Display this help message
    --schema   Custom path to your Prisma schema

${chalk.bold('Commands')}
     pull   Pull the state from the database to the Prisma schema using introspection
     push   Push the state from Prisma schema to the database during prototyping
     seed   Seed your database
  execute   Execute native commands to your database

${chalk.bold('Examples')}

  Run \`prisma db pull\`
  ${chalk.dim('$')} prisma db pull

  Run \`prisma db push\`
  ${chalk.dim('$')} prisma db push

  Run \`prisma db seed\`
  ${chalk.dim('$')} prisma db seed

  Run \`prisma db execute\`
  ${chalk.dim('$')} prisma db execute
`)

  private constructor(private readonly cmds: Commands) {}

  /* eslint-disable-next-line @typescript-eslint/require-await */
  public async parse(argv: string[]): Promise<string | Error> {
    const args = arg(argv, {
      '--help': Boolean,
      '-h': '--help',
      '--preview-feature': Boolean,
      '--telemetry-information': String,
    })

    if (isError(args)) {
      return this.help(args.message)
    }

    // display help for help flag or no subcommand
    if (args._.length === 0 || args['--help']) {
      return this.help()
    }

    // check if we have that subcommand
    const cmd = this.cmds[args._[0]]
    if (cmd) {
      const argsForCmd = args['--preview-feature'] ? [...args._.slice(1), `--preview-feature`] : args._.slice(1)
      return cmd.parse(argsForCmd)
    }

    return unknownCommand(DbCommand.help, args._[0])
  }

  public help(error?: string): string | HelpError {
    if (error) {
      return new HelpError(`\n${chalk.bold.red(`!`)} ${error}\n${DbCommand.help}`)
    }
    return DbCommand.help
  }
}
