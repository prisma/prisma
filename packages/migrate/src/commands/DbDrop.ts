import {
  arg,
  canPrompt,
  checkUnsupportedDataProxy,
  Command,
  dropDatabase,
  format,
  getSchemaDir,
  HelpError,
  isError,
  link,
  loadEnvFile,
} from '@prisma/internals'
import chalk from 'chalk'
import prompt from 'prompts'

import { getDatasourceInfo } from '../utils/ensureDatabaseExists'
import { DbNeedsForceError } from '../utils/errors'
import { PreviewFlagError } from '../utils/flagErrors'
import { getSchemaPathAndPrint } from '../utils/getSchemaPathAndPrint'
import { printDatasource } from '../utils/printDatasource'

export class DbDrop implements Command {
  public static new(): DbDrop {
    return new DbDrop()
  }

  private static help = format(`
${process.platform === 'win32' ? '' : chalk.bold('ðŸ’£  ')}Drop the database

${chalk.bold.yellow('WARNING')} ${chalk.bold(
    `Prisma db drop is currently in Preview (${link('https://pris.ly/d/preview')}).
There may be bugs and it's not recommended to use it in production environments.`,
  )}
${chalk.dim('When using any of the subcommands below you need to explicitly opt-in via the --preview-feature flag.')}

${chalk.bold('Usage')}

  ${chalk.dim('$')} prisma db drop [options] --preview-feature

${chalk.bold('Options')}

   -h, --help   Display this help message
     --schema   Custom path to your Prisma schema
  -f, --force   Skip the confirmation prompt

${chalk.bold('Examples')}

  Drop the database
  ${chalk.dim('$')} prisma db drop --preview-feature

  Specify a schema
  ${chalk.dim('$')} prisma db drop --preview-feature --schema=./schema.prisma

  Use --force to skip the confirmation prompt
  ${chalk.dim('$')} prisma db drop --preview-feature --force
`)

  public async parse(argv: string[]): Promise<string | Error> {
    const args = arg(argv, {
      '--help': Boolean,
      '-h': '--help',
      '--preview-feature': Boolean,
      '--force': Boolean,
      '-f': '--force',
      '--schema': String,
      '--telemetry-information': String,
    })

    if (isError(args)) {
      return this.help(args.message)
    }

    await checkUnsupportedDataProxy('db drop', args, true)

    if (args['--help']) {
      return this.help()
    }

    if (!args['--preview-feature']) {
      throw new PreviewFlagError()
    }

    loadEnvFile(args['--schema'], true)

    const schemaPath = await getSchemaPathAndPrint(args['--schema'])

    const datasourceInfo = await getDatasourceInfo({ schemaPath, throwIfEnvError: true })
    printDatasource({ datasourceInfo })

    const schemaDir = (await getSchemaDir(schemaPath))!

    console.info() // empty line

    if (!args['--force']) {
      if (!canPrompt()) {
        throw new DbNeedsForceError('drop')
      }

      const confirmation = await prompt({
        type: 'text',
        name: 'value',
        message: `Enter the ${datasourceInfo.prettyProvider} database name "${
          datasourceInfo.dbName
        }" to drop it.\nLocation: "${datasourceInfo.dbLocation}".\n${chalk.red('All data will be lost')}.`,
      })
      console.info() // empty line

      if (!confirmation.value) {
        console.info('Drop cancelled.')
        // Return SIGINT exit code to signal that the process was cancelled.
        process.exit(130)
      } else if (confirmation.value !== datasourceInfo.dbName) {
        throw Error(`The database name entered "${confirmation.value}" doesn't match "${datasourceInfo.dbName}".`)
      }
    }

    // Url exists because we set `throwIfEnvErrors: true` in `getDatasourceInfo`
    if (await dropDatabase(datasourceInfo.url!, schemaDir)) {
      return `${process.platform === 'win32' ? '' : 'ðŸš€  '}The ${datasourceInfo.prettyProvider} database "${
        datasourceInfo.dbName
      }" from "${datasourceInfo.dbLocation}" was successfully dropped.\n`
    } else {
      return ''
    }
  }

  public help(error?: string): string | HelpError {
    if (error) {
      return new HelpError(`\n${chalk.bold.red(`!`)} ${error}\n${DbDrop.help}`)
    }
    return DbDrop.help
  }
}
