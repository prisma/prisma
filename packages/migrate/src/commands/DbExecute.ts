import {
  arg,
  checkUnsupportedDataProxy,
  Command,
  format,
  getCommandWithExecutor,
  getSchemaPath,
  HelpError,
  isError,
  loadEnvFile,
  logger,
} from '@prisma/internals'
import chalk from 'chalk'
import fs from 'fs'
import getStdin from 'get-stdin'
import path from 'path'

import { Migrate } from '../Migrate'
import type { EngineArgs } from '../types'

const helpOptions = format(
  `${chalk.bold('Usage')}

${chalk.dim('$')} prisma db execute [options]

${chalk.bold('Options')}

-h, --help            Display this help message

${chalk.italic('Datasource input, only 1 must be provided:')}
--url                 URL of the datasource to run the command on
--schema              Path to your Prisma schema file to take the datasource URL from

${chalk.italic('Script input, only 1 must be provided:')}
--file                Path to a file. The content will be sent as the script to be executed

${chalk.bold('Flags')}

--stdin              Use the terminal standard input as the script to be executed`,
)

export class DbExecute implements Command {
  public static new(): DbExecute {
    return new DbExecute()
  }

  private static help = format(`
${process.platform === 'win32' ? '' : chalk.bold('üìù ')}Execute native commands to your database

This command takes as input a datasource, using ${chalk.green(`--url`)} or ${chalk.green(
    `--schema`,
  )} and a script, using ${chalk.green(`--stdin`)} or ${chalk.green(`--file`)}.
The input parameters are mutually exclusive, only 1 of each (datasource & script) must be provided.
 
The output of the command is connector-specific, and is not meant for returning data, but only to report success or failure.

On SQL databases, this command takes as input a SQL script.
The whole script will be sent as a single command to the database.

${chalk.italic(`This command is currently not supported on MongoDB.`)}

${helpOptions}
${chalk.bold('Examples')}
 
  Execute the content of a SQL script file to the datasource URL taken from the schema
  ${chalk.dim('$')} prisma db execute
    --file ./script.sql \\
    --schema schema.prisma

  Execute the SQL script from stdin to the datasource URL specified via the \`DATABASE_URL\` environment variable
  ${chalk.dim('$')} echo 'TRUNCATE TABLE dev;' | \\
    prisma db execute \\
    --stdin \\
    --url="$DATABASE_URL"

  Like previous example, but exposing the datasource url credentials to your terminal history
  ${chalk.dim('$')} echo 'TRUNCATE TABLE dev;' | \\
    prisma db execute \\
    --stdin \\
    --url="mysql://root:root@localhost/mydb"
`)

  public async parse(argv: string[]): Promise<string | Error> {
    const args = arg(
      argv,
      {
        '--help': Boolean,
        '-h': '--help',
        '--stdin': Boolean,
        '--file': String,
        '--schema': String,
        '--url': String,
        '--preview-feature': Boolean,
        '--telemetry-information': String,
      },
      false,
    )

    if (isError(args)) {
      return this.help(args.message)
    }

    await checkUnsupportedDataProxy('db execute', args, !args['--url'])

    if (args['--help']) {
      return this.help()
    }

    if (args['--preview-feature']) {
      logger.warn(`"prisma db execute" was in Preview and is now Generally Available.
You can now remove the ${chalk.red('--preview-feature')} flag.`)
    }

    loadEnvFile(args['--schema'], false)

    // One of --stdin or --file is required
    if (args['--stdin'] && args['--file']) {
      throw new Error(
        `--stdin and --file cannot be used at the same time. Only 1 must be provided. 
See \`${chalk.green(getCommandWithExecutor('prisma db execute -h'))}\``,
      )
    } else if (!args['--stdin'] && !args['--file']) {
      throw new Error(
        `Either --stdin or --file must be provided.
See \`${chalk.green(getCommandWithExecutor('prisma db execute -h'))}\``,
      )
    }

    // One of --url or --schema is required
    if (args['--url'] && args['--schema']) {
      throw new Error(
        `--url and --schema cannot be used at the same time. Only 1 must be provided.
See \`${chalk.green(getCommandWithExecutor('prisma db execute -h'))}\``,
      )
    } else if (!args['--url'] && !args['--schema']) {
      throw new Error(
        `Either --url or --schema must be provided.
See \`${chalk.green(getCommandWithExecutor('prisma db execute -h'))}\``,
      )
    }

    let script = ''
    // Read file
    if (args['--file']) {
      try {
        script = fs.readFileSync(path.resolve(args['--file']), 'utf-8')
      } catch (e) {
        if (e.code === 'ENOENT') {
          throw new Error(`Provided --file at ${args['--file']} doesn't exist.`)
        } else {
          console.error(`An error occurred while reading the provided --file at ${args['--file']}`)
          throw e
        }
      }
    }
    // Read stdin
    if (args['--stdin']) {
      script = await getStdin()
    }

    let datasourceType: EngineArgs.DbExecuteDatasourceType

    // Execute command(s) to url passed
    if (args['--url']) {
      datasourceType = {
        tag: 'url',
        url: args['--url'],
      }
    }
    // Execute command(s) to url from schema
    else {
      // validate that schema file exists
      // throws an error if it doesn't
      const schemaPath = await getSchemaPath(args['--schema'])

      // Execute command(s) to url from schema
      datasourceType = {
        tag: 'schema',
        // if schemaPath is undefined, getSchemaPath will error
        schema: schemaPath!,
      }
    }

    const migrate = new Migrate()

    try {
      await migrate.engine.dbExecute({
        script,
        datasourceType,
      })
    } finally {
      migrate.stop()
    }

    return `Script executed successfully.`
  }

  public help(error?: string): string | HelpError {
    if (error) {
      throw new HelpError(`\n${error}\n\n${helpOptions}`)
    }
    return DbExecute.help
  }
}
