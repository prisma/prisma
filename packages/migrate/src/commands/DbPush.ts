import {
  arg,
  canPrompt,
  checkUnsupportedDataProxy,
  Command,
  format,
  formatms,
  getCommandWithExecutor,
  HelpError,
  isError,
  loadEnvFile,
  logger,
  protocolToConnectorType,
} from '@prisma/internals'
import chalk from 'chalk'
import prompt from 'prompts'

import { Migrate } from '../Migrate'
import type { EngineResults } from '../types'
import { ensureDatabaseExists, getDatasourceInfo } from '../utils/ensureDatabaseExists'
import { DbPushForceFlagRenamedError, DbPushIgnoreWarningsWithFlagError } from '../utils/errors'
import { getSchemaPathAndPrint } from '../utils/getSchemaPathAndPrint'
import { printDatasource } from '../utils/printDatasource'

export class DbPush implements Command {
  public static new(): DbPush {
    return new DbPush()
  }

  private static help = format(`
${process.platform === 'win32' ? '' : chalk.bold('üôå  ')}Push the state from your Prisma schema to your database

${chalk.bold('Usage')}

  ${chalk.dim('$')} prisma db push [options]

${chalk.bold('Options')}

           -h, --help   Display this help message
             --schema   Custom path to your Prisma schema
   --accept-data-loss   Ignore data loss warnings
        --force-reset   Force a reset of the database before push 
      --skip-generate   Skip triggering generators (e.g. Prisma Client)

${chalk.bold('Examples')}

  Push the Prisma schema state to the database
  ${chalk.dim('$')} prisma db push

  Specify a schema
  ${chalk.dim('$')} prisma db push --schema=./schema.prisma

  Ignore data loss warnings
  ${chalk.dim('$')} prisma db push --accept-data-loss
`)

  public async parse(argv: string[]): Promise<string | Error> {
    const args = arg(
      argv,
      {
        '--help': Boolean,
        '-h': '--help',
        '--preview-feature': Boolean,
        '--accept-data-loss': Boolean,
        '--force-reset': Boolean,
        '--skip-generate': Boolean,
        '--schema': String,
        '--telemetry-information': String,
        // Deprecated
        // --force renamed to --accept-data-loss in 2.17.0
        '--force': Boolean,
        '-f': '--force',
      },
      false,
    )

    if (isError(args)) {
      return this.help(args.message)
    }

    await checkUnsupportedDataProxy('db push', args, true)

    if (args['--help']) {
      return this.help()
    }

    if (args['--preview-feature']) {
      logger.warn(`Prisma "db push" was in Preview and is now Generally Available.
You can now remove the ${chalk.red('--preview-feature')} flag.`)
    }

    if (args['--force']) {
      throw new DbPushForceFlagRenamedError()
    }

    loadEnvFile(args['--schema'], true)

    const schemaPath = await getSchemaPathAndPrint(args['--schema'])

    const datasourceInfo = await getDatasourceInfo({ schemaPath })
    printDatasource({ datasourceInfo })

    const migrate = new Migrate(schemaPath)

    try {
      // Automatically create the database if it doesn't exist
      const wasDbCreated = await ensureDatabaseExists('push', schemaPath)
      if (wasDbCreated) {
        console.info() // empty line
        console.info(wasDbCreated)
      }
    } catch (e) {
      console.info() // empty line
      throw e
    }

    let wasDatabaseReset = false
    if (args['--force-reset']) {
      console.info()
      try {
        await migrate.reset()
      } catch (e) {
        migrate.stop()
        throw e
      }

      let successfulResetMsg = `The ${datasourceInfo.prettyProvider} database`
      if (datasourceInfo.dbName) {
        successfulResetMsg += ` "${datasourceInfo.dbName}"`
      }

      const schemasLength = datasourceInfo.schemas?.length || 0
      // If schemas are defined in the datasource block, print them
      if (datasourceInfo.schemas && schemasLength > 0) {
        successfulResetMsg += ` schema${schemasLength > 1 ? 's' : ''} "${datasourceInfo.schemas.join(', ')}"`
      }
      // Otherwise, print the schema if it's defined in the connection string
      else if (datasourceInfo.schema) {
        successfulResetMsg += ` schema "${datasourceInfo.schema}"`
      }

      if (datasourceInfo.dbLocation) {
        successfulResetMsg += ` at "${datasourceInfo.dbLocation}"`
      }

      successfulResetMsg += ` ${schemasLength > 1 ? 'were' : 'was'} successfully reset.`
      console.info(successfulResetMsg)

      wasDatabaseReset = true
    }

    const before = Date.now()
    let migration: EngineResults.SchemaPush
    try {
      migration = await migrate.push({
        force: args['--accept-data-loss'],
      })
    } catch (e) {
      migrate.stop()
      throw e
    }

    if (migration.unexecutable && migration.unexecutable.length > 0) {
      const messages: string[] = []
      messages.push(`${chalk.bold.red('\n‚ö†Ô∏è We found changes that cannot be executed:\n')}`)
      for (const item of migration.unexecutable) {
        messages.push(`${chalk(`  ‚Ä¢ ${item}`)}`)
      }
      console.info() // empty line

      if (!canPrompt()) {
        migrate.stop()
        throw new Error(`${messages.join('\n')}\n
Use the --force-reset flag to drop the database before push like ${chalk.bold.greenBright(
          getCommandWithExecutor('prisma db push --force-reset'),
        )}
${chalk.bold.redBright('All data will be lost.')}
        `)
      } else {
        console.info(`${messages.join('\n')}\n`)
      }

      console.info() // empty line
      const confirmation = await prompt({
        type: 'confirm',
        name: 'value',
        message: `To apply this change we need to reset the database, do you want to continue? ${chalk.red(
          'All data will be lost',
        )}.`,
      })

      if (!confirmation.value) {
        console.info('Reset cancelled.')
        migrate.stop()
        // Return SIGINT exit code to signal that the process was cancelled.
        process.exit(130)
      }

      try {
        // Reset first to remove all structure and data
        await migrate.reset()
        if (datasourceInfo.dbName && datasourceInfo.dbLocation) {
          console.info(
            `The ${datasourceInfo.prettyProvider} database "${datasourceInfo.dbName}" from "${datasourceInfo.dbLocation}" was successfully reset.`,
          )
        } else {
          console.info(`The ${datasourceInfo.prettyProvider} database was successfully reset.`)
        }
        wasDatabaseReset = true

        // And now we can db push
        await migrate.push({})
      } catch (e) {
        migrate.stop()
        throw e
      }
    }

    if (migration.warnings && migration.warnings.length > 0) {
      console.info(chalk.bold.yellow(`\n‚ö†Ô∏è  There might be data loss when applying the changes:\n`))

      for (const warning of migration.warnings) {
        console.info(chalk(`  ‚Ä¢ ${warning}`))
      }
      console.info() // empty line

      if (!args['--accept-data-loss']) {
        if (!canPrompt()) {
          migrate.stop()
          throw new DbPushIgnoreWarningsWithFlagError()
        }

        console.info() // empty line
        const confirmation = await prompt({
          type: 'confirm',
          name: 'value',
          message: `Do you want to ignore the warning(s)?`,
        })

        if (!confirmation.value) {
          console.info('Push cancelled.')
          migrate.stop()
          // Return SIGINT exit code to signal that the process was cancelled.
          process.exit(130)
        }

        try {
          await migrate.push({
            force: true,
          })
        } catch (e) {
          migrate.stop()
          throw e
        }
      }
    }

    migrate.stop()

    if (!wasDatabaseReset && migration.warnings.length === 0 && migration.executedSteps === 0) {
      console.info(`\nThe database is already in sync with the Prisma schema.`)
    } else {
      const migrationTimeMessage = `Done in ${formatms(Date.now() - before)}`
      const rocketEmoji = process.platform === 'win32' ? '' : 'üöÄ  '
      const migrationSuccessStdMessage = 'Your database is now in sync with your Prisma schema.'
      const migrationSuccessMongoMessage = 'Your database indexes are now in sync with your Prisma schema.'

      // this is safe, as if the protocol was unknown, we would have already exited the program with an error
      const provider = protocolToConnectorType(`${datasourceInfo.url?.split(':')[0]}:`)

      console.info(
        `\n${rocketEmoji}${
          provider === 'mongodb' ? migrationSuccessMongoMessage : migrationSuccessStdMessage
        } ${migrationTimeMessage}`,
      )
    }

    // Run if not skipped
    if (!process.env.PRISMA_MIGRATE_SKIP_GENERATE && !args['--skip-generate']) {
      await migrate.tryToRunGenerate()
    }

    return ``
  }

  public help(error?: string): string | HelpError {
    if (error) {
      return new HelpError(`\n${chalk.bold.red(`!`)} ${error}\n${DbPush.help}`)
    }
    return DbPush.help
  }
}
