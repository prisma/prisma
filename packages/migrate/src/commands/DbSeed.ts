import { arg, Command, format, getSchemaPath, HelpError, isError, loadEnvFile, logger } from '@prisma/internals'
import chalk from 'chalk'

import {
  executeSeedCommand,
  getSeedCommandFromPackageJson,
  legacyTsNodeScriptWarning,
  verifySeedConfigAndReturnMessage,
} from '../utils/seed'

export class DbSeed implements Command {
  public static new(): DbSeed {
    return new DbSeed()
  }

  private static help = format(`
${process.platform === 'win32' ? '' : chalk.bold('ðŸ™Œ  ')}Seed your database

${chalk.bold('Usage')}

  ${chalk.dim('$')} prisma db seed [options]

${chalk.bold('Options')}

  -h, --help   Display this help message
`)

  public async parse(argv: string[]): Promise<string | Error> {
    const args = arg(
      argv,
      {
        '--help': Boolean,
        '-h': '--help',
        '--preview-feature': Boolean,
        '--schema': String,
        '--telemetry-information': String,
      },
      false,
    )

    if (isError(args)) {
      return this.help(args.message)
    }

    if (args['--help']) {
      return this.help()
    }

    if (args['--preview-feature']) {
      logger.warn(`Prisma "db seed" was in Preview and is now Generally Available.
You can now remove the ${chalk.red('--preview-feature')} flag.`)

      // Print warning if user has a "ts-node" script in their package.json, not supported anymore
      await legacyTsNodeScriptWarning()
    }

    loadEnvFile(args['--schema'], true)

    // Print warning if user is using --schema
    if (args['--schema']) {
      logger.warn(
        chalk.yellow(
          `The "--schema" parameter is not used anymore by "prisma db seed" since version 3.0 and can now be removed.`,
        ),
      )
    }

    const seedCommandFromPkgJson = await getSeedCommandFromPackageJson(process.cwd())

    if (!seedCommandFromPkgJson) {
      // Only used to help users to set up their seeds from old way to new package.json config
      const schemaPath = await getSchemaPath(args['--schema'])

      const message = await verifySeedConfigAndReturnMessage(schemaPath)
      // Error because setup of the feature needs to be done
      if (message) {
        throw new Error(message)
      }

      return ``
    }

    // Seed command is set
    // Execute user seed command
    const successfulSeeding = await executeSeedCommand(seedCommandFromPkgJson)
    if (successfulSeeding) {
      return `\n${process.platform === 'win32' ? '' : 'ðŸŒ±  '}The seed command has been executed.`
    } else {
      process.exit(1)
    }
  }

  public help(error?: string): string | HelpError {
    if (error) {
      return new HelpError(`\n${chalk.bold.red(`!`)} ${error}\n${DbSeed.help}`)
    }
    return DbSeed.help
  }
}
