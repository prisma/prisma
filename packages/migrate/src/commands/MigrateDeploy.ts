import Debug from '@prisma/debug'
import { arg, checkUnsupportedDataProxy, Command, format, HelpError, isError, loadEnvFile } from '@prisma/internals'
import chalk from 'chalk'

import { Migrate } from '../Migrate'
import { throwUpgradeErrorIfOldMigrate } from '../utils/detectOldMigrate'
import { ensureDatabaseExists, getDatasourceInfo } from '../utils/ensureDatabaseExists'
import { EarlyAccessFeatureFlagWithMigrateError, ExperimentalFlagWithMigrateError } from '../utils/flagErrors'
import { getSchemaPathAndPrint } from '../utils/getSchemaPathAndPrint'
import { printDatasource } from '../utils/printDatasource'
import { printFilesFromMigrationIds } from '../utils/printFiles'

const debug = Debug('prisma:migrate:deploy')

export class MigrateDeploy implements Command {
  public static new(): MigrateDeploy {
    return new MigrateDeploy()
  }

  private static help = format(`
Apply pending migrations to update the database schema in production/staging

${chalk.bold('Usage')}

  ${chalk.dim('$')} prisma migrate deploy [options]

${chalk.bold('Options')}

  -h, --help   Display this help message
    --schema   Custom path to your Prisma schema

${chalk.bold('Examples')}

  Deploy your pending migrations to your production/staging database
  ${chalk.dim('$')} prisma migrate deploy

  Specify a schema
  ${chalk.dim('$')} prisma migrate deploy --schema=./schema.prisma

`)

  public async parse(argv: string[]): Promise<string | Error> {
    const args = arg(
      argv,
      {
        '--help': Boolean,
        '-h': '--help',
        '--experimental': Boolean,
        '--early-access-feature': Boolean,
        '--schema': String,
        '--telemetry-information': String,
      },
      false,
    )

    if (isError(args)) {
      return this.help(args.message)
    }

    await checkUnsupportedDataProxy('migrate deploy', args, true)

    if (args['--help']) {
      return this.help()
    }

    if (args['--experimental']) {
      throw new ExperimentalFlagWithMigrateError()
    }

    if (args['--early-access-feature']) {
      throw new EarlyAccessFeatureFlagWithMigrateError()
    }

    loadEnvFile(args['--schema'], true)

    const schemaPath = await getSchemaPathAndPrint(args['--schema'])

    printDatasource({ datasourceInfo: await getDatasourceInfo({ schemaPath }) })

    throwUpgradeErrorIfOldMigrate(schemaPath)

    const migrate = new Migrate(schemaPath)

    try {
      // Automatically create the database if it doesn't exist
      const wasDbCreated = await ensureDatabaseExists('apply', schemaPath)
      if (wasDbCreated) {
        console.info() // empty line
        console.info(wasDbCreated)
      }
    } catch (e) {
      console.info() // empty line
      throw e
    }

    const listMigrationDirectoriesResult = await migrate.listMigrationDirectories()
    debug({ listMigrationDirectoriesResult })

    console.info() // empty line
    if (listMigrationDirectoriesResult.migrations.length > 0) {
      const migrations = listMigrationDirectoriesResult.migrations
      console.info(`${migrations.length} migration${migrations.length > 1 ? 's' : ''} found in prisma/migrations`)
    } else {
      console.info(`No migration found in prisma/migrations`)
    }

    let migrationIds: string[]
    try {
      console.info() // empty line
      const { appliedMigrationNames } = await migrate.applyMigrations()
      migrationIds = appliedMigrationNames
    } finally {
      // Stop engine
      migrate.stop()
    }

    console.info() // empty line
    if (migrationIds.length === 0) {
      return chalk.greenBright(`No pending migrations to apply.`)
    } else {
      return `The following migration${migrationIds.length > 1 ? 's' : ''} have been applied:\n\n${chalk(
        printFilesFromMigrationIds('migrations', migrationIds, {
          'migration.sql': '',
        }),
      )}
      
${chalk.greenBright('All migrations have been successfully applied.')}`
    }
  }

  public help(error?: string): string | HelpError {
    if (error) {
      return new HelpError(`\n${chalk.bold.red(`!`)} ${error}\n${MigrateDeploy.help}`)
    }
    return MigrateDeploy.help
  }
}
