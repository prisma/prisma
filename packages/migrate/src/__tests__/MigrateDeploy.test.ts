import { jestConsoleContext, jestContext } from '@prisma/internals'
import fs from 'fs-jetpack'

import { MigrateDeploy } from '../commands/MigrateDeploy'

const ctx = jestContext.new().add(jestConsoleContext()).assemble()

describe('common', () => {
  it('should fail if no schema file', async () => {
    ctx.fixture('empty')
    const result = MigrateDeploy.new().parse([])
    await expect(result).rejects.toThrowErrorMatchingInlineSnapshot(`
            Could not find a schema.prisma file that is required for this command.
            You can either provide it with --schema, set it as \`prisma.schema\` in your package.json or put it into the default location ./prisma/schema.prisma https://pris.ly/d/prisma-schema-location
          `)
  })
  it('should fail if experimental flag', async () => {
    ctx.fixture('empty')
    const result = MigrateDeploy.new().parse(['--experimental'])
    await expect(result).rejects.toThrowErrorMatchingInlineSnapshot(`
            Prisma Migrate was Experimental and is now Generally Available.
            WARNING this new version has some breaking changes to use it it's recommended to read the documentation first and remove the --experimental flag.
          `)
  })
  it('should fail if early access flag', async () => {
    ctx.fixture('empty')
    const result = MigrateDeploy.new().parse(['--early-access-feature'])
    await expect(result).rejects.toThrowErrorMatchingInlineSnapshot(`
            Prisma Migrate was in Early Access and is now Generally Available.
            Remove the --early-access-feature flag.
          `)
  })
})

describe('sqlite', () => {
  it('no unapplied migrations', async () => {
    ctx.fixture('schema-only-sqlite')
    const result = MigrateDeploy.new().parse(['--schema=./prisma/empty.prisma'])
    await expect(result).resolves.toMatchInlineSnapshot(`No pending migrations to apply.`)

    expect(ctx.mocked['console.info'].mock.calls.join('\n')).toMatchInlineSnapshot(`
      Prisma schema loaded from prisma/empty.prisma
      Datasource "my_db": SQLite database "dev.db" at "file:dev.db"

      SQLite database dev.db created at file:dev.db

      No migration found in prisma/migrations


    `)
    expect(ctx.mocked['console.log'].mock.calls).toMatchSnapshot()
    expect(ctx.mocked['console.error'].mock.calls).toMatchSnapshot()
  })

  it('1 unapplied migration', async () => {
    ctx.fixture('existing-db-1-migration')
    fs.remove('prisma/dev.db')

    const result = MigrateDeploy.new().parse([])
    await expect(result).resolves.toMatchInlineSnapshot(`
            The following migration have been applied:

            migrations/
              └─ 20201231000000_init/
                └─ migration.sql
                  
            All migrations have been successfully applied.
          `)

    // Second time should do nothing (already applied)
    const resultBis = MigrateDeploy.new().parse([])
    await expect(resultBis).resolves.toMatchInlineSnapshot(`No pending migrations to apply.`)

    expect(ctx.mocked['console.info'].mock.calls.join('\n')).toMatchInlineSnapshot(`
      Prisma schema loaded from prisma/schema.prisma
      Datasource "my_db": SQLite database "dev.db" at "file:dev.db"

      SQLite database dev.db created at file:dev.db

      1 migration found in prisma/migrations

      Applying migration \`20201231000000_init\`

      Prisma schema loaded from prisma/schema.prisma
      Datasource "my_db": SQLite database "dev.db" at "file:dev.db"

      1 migration found in prisma/migrations


    `)
    expect(ctx.mocked['console.log'].mock.calls).toMatchSnapshot()
    expect(ctx.mocked['console.error'].mock.calls).toMatchSnapshot()
  })

  it('should throw if database is not empty', async () => {
    ctx.fixture('existing-db-1-migration-conflict')

    const result = MigrateDeploy.new().parse([])
    await expect(result).rejects.toMatchInlineSnapshot(`
            P3005

            The database schema is not empty. Read more about how to baseline an existing production database: https://pris.ly/d/migrate-baseline

          `)

    expect(ctx.mocked['console.info'].mock.calls.join('\n')).toMatchInlineSnapshot(`
      Prisma schema loaded from prisma/schema.prisma
      Datasource "my_db": SQLite database "dev.db" at "file:dev.db"

      1 migration found in prisma/migrations

    `)
    expect(ctx.mocked['console.log'].mock.calls).toMatchSnapshot()
    expect(ctx.mocked['console.error'].mock.calls).toMatchSnapshot()
  })
})
