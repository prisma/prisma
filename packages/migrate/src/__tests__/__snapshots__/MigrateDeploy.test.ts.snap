// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sqlite 1 unapplied migration 4`] = `Array []`;

exports[`sqlite 1 unapplied migration 5`] = `Array []`;

exports[`sqlite no unapplied migrations 3`] = `Array []`;

exports[`sqlite no unapplied migrations 4`] = `Array []`;

exports[`sqlite should throw if database is not empty 3`] = `Array []`;

exports[`sqlite should throw if database is not empty 4`] = `Array []`;
